app/uk/gov/hmrc/gform/models/Bracket.scala

```diff
@@ -0,0 +1,191 @@
+
+  def toPageModel: NonEmptyList[PageModel[A]] =
+    fold[NonEmptyList[PageModel[A]]](a => NonEmptyList.one(a.singleton))(_.singletons.map(_.singleton))(
+      _.iterations.flatMap(_.toPageModel))
+
+  def filter(predicate: PageModel[A] => Boolean): Option[Bracket[A]] = this match {
+    case Bracket.AddToList(iterations, source) =>
+      val filtered: Option[NonEmptyList[Bracket.AddToListIteration[A]]] = iterations.traverse(_.filter(predicate))
+      filtered.map(Bracket.AddToList(_, source))
+    case Bracket.RepeatingPage(singletons, source) =>
+      val filtered = singletons.filter(singletonWithNumber => predicate(singletonWithNumber.singleton))
+      NonEmptyList.fromList(filtered).map(Bracket.RepeatingPage(_, source))
+    case b @ Bracket.NonRepeatingPage(singleton, _, _) => if (predicate(singleton)) Some(b) else None
+  }
+
+}
+
+object Bracket {
+  case class AddToListIteration[A <: PageMode](
+    singletons: NonEmptyList[SingletonWithNumber[A]], // There must be at least one page in Add-to-list iteration
+    repeater: RepeaterWithNumber[A]
+  ) {
+
+    def toPageModelWithNumber: NonEmptyList[(PageModel[A], SectionNumber)] =
+      singletons.map(_.toPageModelWithNumber) ::: NonEmptyList.one(repeater.repeater -> repeater.sectionNumber)
+
+    def toPlainBracket: BracketPlain.AddToListIteration[A] =
+      BracketPlain.AddToListIteration(singletons.map(_.singleton), repeater.repeater)
+    def hasSectionNumber(sectionNumber: SectionNumber): Boolean =
+      repeater.sectionNumber === sectionNumber || singletons.exists(_.sectionNumber === sectionNumber)
+    def singleton(sectionNumber: SectionNumber): Singleton[A] =
+      singletons.toList
+        .collectFirst {
+          case singletonWithNumber if singletonWithNumber.sectionNumber === sectionNumber =>
+            singletonWithNumber.singleton
+        }
+        .getOrElse(throw new IllegalArgumentException(s"Invalid sectionNumber: $sectionNumber for AddToListIteration"))
+
+    def toPageModel: NonEmptyList[PageModel[A]] = singletons.map(_.singleton) ::: NonEmptyList.one(repeater.repeater)
+
+    def map[B <: PageMode](f: Singleton[A] => Singleton[B], g: Repeater[A] => Repeater[B]): AddToListIteration[B] =
+      AddToListIteration(singletons.map(_.map(f)), repeater.map(g))
+
+    def filter(predicate: PageModel[A] => Boolean): Option[AddToListIteration[A]] = {
+      val filtered = singletons.filter(s => predicate(s.singleton))
+      NonEmptyList.fromList(filtered).map(AddToListIteration(_, repeater))
+    }
+
+    def firstSectionNumber: SectionNumber = singletons.head.sectionNumber
+
+  }
+
+  case class NonRepeatingPage[A <: PageMode](
+    singleton: Singleton[A],
+    sectionNumber: SectionNumber,
+    source: Section.NonRepeatingPage
+  ) extends Bracket[A] {
+    def map[B <: PageMode](f: Singleton[A] => Singleton[B], g: Repeater[A] => Repeater[B]): NonRepeatingPage[B] =
```

Comment:
`g: Repeater[A] => Repeater[B]` seems unused. is it required?

Reply:
`map` is `override` method.